internal/servers/plugin/v3/plugin.go

```diff
@@ -0,0 +1,273 @@
+	return &pb.GetVersion_Response{
+		Version: s.Plugin.Version(),
+	}, nil
+}
+
+func (s *Server) Init(ctx context.Context, req *pb.Init_Request) (*pb.Init_Response, error) {
+	if err := s.Plugin.Init(ctx, req.Spec); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to init plugin: %v", err)
+	}
+	return &pb.Init_Response{}, nil
+}
+
+func (s *Server) Sync(req *pb.Sync_Request, stream pb.Plugin_SyncServer) error {
+	msgs := make(chan message.Message)
+	var syncErr error
+	ctx := stream.Context()
+
+	syncOptions := plugin.SyncOptions{
+		Tables:     req.Tables,
+		SkipTables: req.SkipTables,
+	}
+
+	if req.StateBackend != nil {
+		opts := []managedplugin.Option{
+			managedplugin.WithLogger(s.Logger),
+			managedplugin.WithDirectory(s.Directory),
+		}
+		if s.NoSentry {
+			opts = append(opts, managedplugin.WithNoSentry())
+		}
+		statePlugin, err := managedplugin.NewClient(ctx, managedplugin.PluginDestination, managedplugin.Config{
+			Path:     req.StateBackend.Path,
+			Registry: managedplugin.Registry(req.StateBackend.Registry),
+			Version:  req.StateBackend.Version,
+		}, opts...)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to create state plugin: %v", err)
+		}
+		stateClient, err := newStateClient(ctx, statePlugin.Conn, req.StateBackend)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to create state client: %v", err)
+		}
+		syncOptions.StateBackend = stateClient
+	}
+
+	go func() {
+		defer close(msgs)
+		err := s.Plugin.Sync(ctx, syncOptions, msgs)
+		if err != nil {
+			syncErr = fmt.Errorf("failed to sync records: %w", err)
+		}
+	}()
+
+	pbMsg := &pb.Sync_Response{}
+	for msg := range msgs {
+		switch m := msg.(type) {
+		case *message.MigrateTable:
+			tableSchema := m.Table.ToArrowSchema()
+			schemaBytes, err := schema.ToBytes(tableSchema)
+			if err != nil {
+				return status.Errorf(codes.Internal, "failed to encode table schema: %v", err)
+			}
+			pbMsg.Message = &pb.Sync_Response_MigrateTable{
+				MigrateTable: &pb.MessageMigrateTable{
+					Table: schemaBytes,
+				},
+			}
+
+		case *message.Insert:
+			recordBytes, err := schema.RecordToBytes(m.Record)
+			if err != nil {
+				return status.Errorf(codes.Internal, "failed to encode record: %v", err)
+			}
+			pbMsg.Message = &pb.Sync_Response_Insert{
+				Insert: &pb.MessageInsert{
+					Record: recordBytes,
+				},
+			}
+		case *message.DeleteStale:
+			tableBytes, err := m.Table.ToArrowSchemaBytes()
+			if err != nil {
+				return status.Errorf(codes.Internal, "failed to encode record: %v", err)
+			}
+			pbMsg.Message = &pb.Sync_Response_Delete{
+				Delete: &pb.MessageDeleteStale{
+					Table:      tableBytes,
+					SourceName: m.SourceName,
+					SyncTime:   timestamppb.New(m.SyncTime),
+				},
+			}
+		default:
+			return status.Errorf(codes.Internal, "unknown message type: %T", msg)
+		}
+
+		// err := checkMessageSize(msg, rec)
```

Comment:
We need to remember to add this back in so we don't have a regression (adding a test would be good too)